### PR TITLE
Adding AOCC compiler to SPACK community

### PR DESCRIPTION
**Pull request by GaneshPrasadMA**
The AOCC compiler system offers a high level of advanced optimizations, multi-threading and processor support that includes global optimization, vectorization, inter-procedural analyses, loop transformations, and code generation. AMD also provides highly optimized libraries, which extract the optimal performance from each x86 processor core when utilized. The AOCC Compiler Suite simplifies and accelerates development and tuning for x86 applications.

### CODE REVIEW

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Needs unit tests for detection and flags, see `lib/spack/spack/test/compilers/*.py`

**Comment by scheibelp** [comment]
https://github.com/spack/spack/pull/19294 has a good example of what tests would be useful to add: see specifically the modifications to `test/compilers/detection.py` and `test/compilers/basics.py`: it would be great to add `test_aocc_version_detection`/`test_aocc_flags ` functions analogous to `test_nvhpc_version_detection`/`test_nvhpc_flags` (the first test asserts an extracted version from the compiler's `--version` output, and the latter asserts the proper flag formats).

I assume this would be sufficient to address @adamjstewart's request (but Adam I think it would be good to state if you have additional tests in mind).

Let me know if it would be problematic to add that here or if you have questions about how those functions

**Comment by GaneshPrasadMA** [comment]
> Needs unit tests for detection and flags, see `lib/spack/spack/test/compilers/*.py`

I have addressed your comments by adding tests in the following files, please review  

Files:
lib/spack/spack/test/compilers/basics.py
lib/spack/spack/test/compilers/detection.py

File: lib/spack/spack/test/compilers/detection.py [hunk 1]
```diff
@@ -233,3 +234,30 @@ def _module(cmd, *args):
...
+     '(based on LLVM Mirror.Version.10.0.0)\n'
+     'Target: x86_64-unknown-linux-gnu\n'
+     'Thread model: posix\n', '2.2.0'
+     ),
+    # Fortran compiler
+    ('AMD clang version 10.0.0 (CLANG: AOCC_2.2.0-Build#93 2020_06_25)'
+     '(based on LLVM Mirror.Version.10.0.0)\n'
+     'Target: x86_64-unknown-linux-gnu\n'
+     'Thread model: posix\n', '2.2.0'
+     )
```
Comment (adamjstewart) [inline lib/spack/spack/test/compilers/detection.py:245]: Are these all the same? If so, there's no need to repeat them, we would just be testing the same thing 3 times.

File: var/spack/repos/builtin/packages/aocc/package.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    family = 'compiler'
+    homepage = "https://developer.amd.com/amd-aocc/"
+    key_part_1 = '500940ce36c19297dfba3aa56dcef33b'
+    key_part_2 = '6145867a1f34890945172ac2be83b286'
+
+    site_nm = 'developer.amd.com'
+    tar_nm = 'aocc-compiler-2.2.0.tar'
+    version(ver="2.2.0",
+            sha256='{0}{1}'.format(key_part_1, key_part_2),
+            url='http://{0}/wordpress/media/files/{1}'.format(site_nm, tar_nm))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/aocc/package.py]: Spack ignores URLs and checksums in its flake8 line-length checks, so you can simplify this quite a bit.

File: lib/spack/spack/test/compilers/detection.py (hunk 1 above)
Comment (GaneshPrasadMA) [reply lib/spack/spack/test/compilers/detection.py:245]: Yes, I have modified as per your review comments

File: var/spack/repos/builtin/packages/aocc/package.py (hunk 2 above)
Comment (GaneshPrasadMA) [reply var/spack/repos/builtin/packages/aocc/package.py]: Yes, I have simplified them, please review

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**